File: sdks/go/feature_flags_test.go
// Tests de feature flags del SDK Go. Cubren:
//
//  1. Vectores dorados del hash sticky (FNV-1a 32-bit), idénticos a Node/Next.
//  2. Flag con rollout=100 → IsFeatureEnabled true + se encola $feature_exposure
//     con variant "B".
//  3. Flag con conditions.properties no satisfechas → false y sin exposición.
//
// httptest.NewServer sirve /api/v1/ingest/feature-flags y /api/v1/ingest/events.
package faro

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

// ---- 1. vectores dorados del hash ----

func TestStickyBucketVectoresDorados(t *testing.T) {
	cases := []struct {
		in   string
		want int
	}{
		{"proj:new-checkout:user_42", 9},
		{"acme:flag-a:anon_x", 54},
		{"myproj:dark-mode:user_1", 75},
		{"p:k:abcdefghij", 49},
		{"demo:exp1:user_42", 34},
	}
	for _, c := range cases {
		if got := stickyBucket(c.in); got != c.want {
			t.Errorf("stickyBucket(%q) = %d, want %d", c.in, got, c.want)
		}
	}
}

// flagServer sirve el endpoint de feature flags con una respuesta fija y
// captura los POST a /ingest/events para inspeccionar exposiciones.
type flagServer struct {
	cap     *captureServer
	flagsJS string
}

func newFlagServer(flagsJS string) *flagServer {
	return &flagServer{cap: newCaptureServer(), flagsJS: flagsJS}
}

func (f *flagServer) handler() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/api/v1/ingest/feature-flags" {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusOK)
			_, _ = io.WriteString(w, f.flagsJS)
			return
		}
		// El resto (events/logs) lo maneja el captureServer.
		f.cap.handler().ServeHTTP(w, r)
	})
}

// ---- 2. rollout=100 → habilitada + exposición variant B ----

func TestIsFeatureEnabledRollout100EmiteExposicion(t *testing.T) {
	fs := newFlagServer(`{
		"project": "demo",
		"flags": [
			{"key": "new-ui", "rollout_percentage": 100, "conditions": {}}
		]
	}`)
	srv := httptest.NewServer(fs.handler())
	defer srv.Close()

	c, err := New(Options{
		Endpoint:                   srv.URL,
		Token:                      "tk",
		Service:                    "ff-test",
		FlushInterval:              50 * time.Millisecond,
		FeatureFlagRefreshInterval: 30 * time.Second, // sin auto-tick durante el test
		OnInternalError:            func(error) {},
	})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close(context.Background())

	// Carga las flags de forma síncrona (sin esperar al ticker).
	c.RefreshFeatureFlags(context.Background())

	if !c.IsFeatureEnabled("new-ui", FlagContext{DistinctID: "user_1"}) {
		t.Fatalf("new-ui con rollout=100 debería estar habilitada")
	}
	// Una flag inexistente siempre es false.
	if c.IsFeatureEnabled("no-existe", FlagContext{DistinctID: "user_1"}) {
		t.Fatalf("flag inexistente debe ser false")
	}

	if err := c.Flush(2 * time.Second); err != nil {
		t.Fatal(err)
	}
	waitFor(t, 2*time.Second, func() bool {
		return len(featureExposureEvents(fs.cap.snapshot())) >= 1
	})

	exposures := featureExposureEvents(fs.cap.snapshot())
	if len(exposures) != 1 {
		t.Fatalf("esperaba exactamente 1 $feature_exposure; got %d (%+v)", len(exposures), exposures)
	}
	e := exposures[0]
	if e["distinct_id"] != "user_1" {
		t.Errorf("distinct_id override = %v, want user_1", e["distinct_id"])
	}
	props := e["properties"].(map[string]any)
	if props["flag_key"] != "new-ui" {
		t.Errorf("flag_key = %v, want new-ui", props["flag_key"])
	}
	if props["variant"] != "B" {
		t.Errorf("variant = %v, want B", props["variant"])
	}
	if props["enabled"] != true {
		t.Errorf("enabled = %v, want true", props["enabled"])
	}

	// Dedup: una segunda evaluación de la misma combinación no debe emitir otro evento.
	c.IsFeatureEnabled("new-ui", FlagContext{DistinctID: "user_1"})
	_ = c.Flush(time.Second)
	if got := len(featureExposureEvents(fs.cap.snapshot())); got != 1 {
		t.Errorf("la exposición debe dedup-arse; got %d eventos", got)
	}
}

// ---- 3. conditions.properties no satisfechas → false sin exposición ----

func TestIsFeatureEnabledCondicionesNoSatisfechas(t *testing.T) {
	fs := newFlagServer(`{
		"project": "demo",
		"flags": [
			{"key": "beta", "rollout_percentage": 100, "conditions": {"properties": {"plan": "pro"}}}
		]
	}`)
	srv := httptest.NewServer(fs.handler())
	defer srv.Close()

	c, err := New(Options{
		Endpoint:                   srv.URL,
		Token:                      "tk",
		Service:                    "ff-cond-test",
		FlushInterval:              50 * time.Millisecond,
		FeatureFlagRefreshInterval: 30 * time.Second,
		OnInternalError:            func(error) {},
	})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close(context.Background())

	c.RefreshFeatureFlags(context.Background())

	// plan != pro → no cumple condiciones → false, sin exposición.
	if c.IsFeatureEnabled("beta", FlagContext{DistinctID: "user_2", Properties: map[string]any{"plan": "free"}}) {
		t.Fatalf("beta no debe activarse con plan=free")
	}
	// Sin la propiedad → tampoco cumple.
	if c.IsFeatureEnabled("beta", FlagContext{DistinctID: "user_3"}) {
		t.Fatalf("beta no debe activarse sin plan")
	}
	// Con plan=pro → cumple condiciones y rollout=100 → true.
	if !c.IsFeatureEnabled("beta", FlagContext{DistinctID: "user_4", Properties: map[string]any{"plan": "pro"}}) {
		t.Fatalf("beta debe activarse con plan=pro")
	}

	_ = c.Flush(2 * time.Second)
	waitFor(t, 2*time.Second, func() bool {
		return len(featureExposureEvents(fs.cap.snapshot())) >= 1
	})

	exposures := featureExposureEvents(fs.cap.snapshot())
	// Solo user_4 (plan=pro) genera exposición; user_2/user_3 no.
	if len(exposures) != 1 {
		t.Fatalf("esperaba 1 exposición (solo plan=pro); got %d (%+v)", len(exposures), exposures)
	}
	if exposures[0]["distinct_id"] != "user_4" {
		t.Errorf("la única exposición debe ser de user_4; got %v", exposures[0]["distinct_id"])
	}
	if exposures[0]["properties"].(map[string]any)["variant"] != "B" {
		t.Errorf("variant esperado B")
	}
}

// ---- helpers ----

// featureExposureEvents extrae los eventos $feature_exposure de los batches.
func featureExposureEvents(batches []map[string]any) []map[string]any {
	var out []map[string]any
	for _, e := range eventsFromCapture(batches) {
		if e["name"] == "$feature_exposure" {
			out = append(out, e)
		}
	}
	return out
}

// Sanity: el body de feature-flags se decodifica como esperamos.
func TestFeatureFlagsResponseDecode(t *testing.T) {
	var body featureFlagsResponse
	js := `{"project":"p","flags":[{"key":"k","rollout_percentage":42,"conditions":{"properties":{"n":1}}}]}`
	if err := json.Unmarshal([]byte(js), &body); err != nil {
		t.Fatal(err)
	}
	if body.Project != "p" || len(body.Flags) != 1 {
		t.Fatalf("decode inesperado: %+v", body)
	}
	if body.Flags[0].Rollout != 42 {
		t.Errorf("rollout = %d, want 42", body.Flags[0].Rollout)
	}
}

File: sdks/flutter/test/feature_flags_test.dart
/// Tests de feature flags + cierre acotado del SDK Flutter.
///
///   (a) Los 5 vectores dorados del sticky bucket (FNV-1a 32-bit) deben coincidir
///       byte a byte con el resto de SDKs (Node es la referencia).
///   (b) /api/v1/ingest/feature-flags con flag rollout=100 → isFeatureEnabled true
///       y se encola un `$feature_exposure` con variant 'B'.
///   (c) Flag con conditions.properties no satisfechas → false y SIN exposición.
///   (d) close(timeout: ...) retorna aunque la red cuelgue (no bloquea).
///
/// El sticky bucket no se expone públicamente, así que (a) se valida de forma
/// equivalente: configuramos un flag con `project`/`key`/`distinct_id` cuyo bucket
/// conocido cae dentro/fuera de un rollout dado, y comprobamos el booleano. Para
/// el resto levantamos un HttpServer local (dart:io) que sirve el JSON de flags.
library;

import 'dart:convert';
import 'dart:io';

import 'package:faro_sdk/faro_sdk.dart';
import 'package:flutter_test/flutter_test.dart';

/// Server local que distingue rutas:
///  - GET  /api/v1/ingest/feature-flags → devuelve `flagsBody` (JSON de flags)
///  - POST /api/v1/ingest/events|logs  → captura y responde 200
/// Permite además colgar (`hang`) para simular red caída.
class _FlagsServer {
  late HttpServer _server;
  final List<Map<String, dynamic>> events = [];
  Map<String, dynamic> flagsBody = {'project': '', 'flags': <dynamic>[]};
  bool hang = false;

  Future<void> start() async {
    _server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
    _server.listen((HttpRequest req) async {
      if (hang) {
        // Nunca responde: simula red colgada (el cliente depende de su timeout).
        return;
      }
      final body = await utf8.decoder.bind(req).join();
      if (req.uri.path == '/api/v1/ingest/feature-flags' &&
          req.method == 'GET') {
        req.response.statusCode = 200;
        req.response.headers.contentType = ContentType.json;
        req.response.write(jsonEncode(flagsBody));
        await req.response.close();
        return;
      }
      // logs/events ingest
      Map<String, dynamic>? parsed;
      try {
        parsed = jsonDecode(body) as Map<String, dynamic>;
      } catch (_) {}
      if (parsed != null) {
        for (final e
            in (parsed['events'] as List<dynamic>? ?? const <dynamic>[])) {
          events.add(e as Map<String, dynamic>);
        }
      }
      req.response.statusCode = 200;
      req.response.headers.contentType = ContentType.json;
      req.response.write('{"ok":true}');
      await req.response.close();
    });
  }

  String get url => 'http://127.0.0.1:${_server.port}';
  Future<void> stop() => _server.close(force: true);
}

Future<void> _waitFor(
  bool Function() cond, {
  Duration timeout = const Duration(seconds: 2),
}) async {
  final deadline = DateTime.now().add(timeout);
  while (DateTime.now().isBefore(deadline)) {
    if (cond()) return;
    await Future<void>.delayed(const Duration(milliseconds: 20));
  }
}

void main() {
  late _FlagsServer server;

  setUp(() async {
    server = _FlagsServer();
    await server.start();
  });

  tearDown(() async {
    try {
      await Faro.instance.close();
    } catch (_) {}
    await server.stop();
  });

  // ---------- (a) Vectores dorados del sticky bucket ----------
  //
  // No exponemos _stickyBucket; lo verificamos vía isFeatureEnabled con un
  // flag cuyo rollout está justo por encima del bucket esperado (→ enabled) y
  // justo en el bucket (→ disabled, porque la condición es bucket < rollout).
  //   "proj:new-checkout:user_42" -> 9
  //   "acme:flag-a:anon_x"        -> 54
  //   "myproj:dark-mode:user_1"   -> 75
  //   "p:k:abcdefghij"            -> 49
  //   "demo:exp1:user_42"         -> 34
  group('sticky bucket — vectores dorados', () {
    final vectors = <List<Object>>[
      ['proj', 'new-checkout', 'user_42', 9],
      ['acme', 'flag-a', 'anon_x', 54],
      ['myproj', 'dark-mode', 'user_1', 75],
      ['p', 'k', 'abcdefghij', 49],
      ['demo', 'exp1', 'user_42', 34],
    ];

    for (final v in vectors) {
      final project = v[0] as String;
      final key = v[1] as String;
      final id = v[2] as String;
      final bucket = v[3] as int;
      test('"$project:$key:$id" -> $bucket', () async {
        server.flagsBody = {
          'project': project,
          'flags': [
            {
              'key': key,
              'rollout_percentage': 100,
              'conditions': <String, Object?>{},
            },
          ],
        };
        final faro = Faro.init(
          FaroOptions(
            endpoint: server.url,
            token: 'tk',
            service: 'ff-golden',
            installGlobalHandlers: false,
            flushInterval: const Duration(seconds: 100),
            featureFlagRefreshInterval: const Duration(seconds: 100),
          ),
        );
        await faro.refreshFeatureFlags();

        // rollout = bucket+1 → bucket < rollout → enabled.
        server.flagsBody = {
          'project': project,
          'flags': [
            {
              'key': key,
              'rollout_percentage': bucket + 1,
              'conditions': <String, Object?>{},
            },
          ],
        };
        await faro.refreshFeatureFlags();
        expect(
          faro.isFeatureEnabled(key, distinctId: id),
          isTrue,
          reason: 'bucket esperado=$bucket, rollout=${bucket + 1} → enabled',
        );

        // rollout = bucket → bucket < rollout es falso → disabled.
        server.flagsBody = {
          'project': project,
          'flags': [
            {
              'key': key,
              'rollout_percentage': bucket,
              'conditions': <String, Object?>{},
            },
          ],
        };
        await faro.refreshFeatureFlags();
        expect(
          faro.isFeatureEnabled(key, distinctId: id),
          isFalse,
          reason: 'bucket esperado=$bucket, rollout=$bucket → disabled',
        );
        await faro.close();
      });
    }
  });

  // ---------- (b) rollout=100 → true + $feature_exposure variant B ----------

  test('rollout=100 → isFeatureEnabled true y encola \$feature_exposure (B)',
      () async {
    server.flagsBody = {
      'project': 'demo',
      'flags': [
        {
          'key': 'new-ui',
          'rollout_percentage': 100,
          'conditions': <String, Object?>{},
        },
      ],
    };
    final faro = Faro.init(
      FaroOptions(
        endpoint: server.url,
        token: 'tk',
        service: 'ff-rollout',
        installGlobalHandlers: false,
        flushInterval: const Duration(seconds: 100),
        featureFlagRefreshInterval: const Duration(seconds: 100),
      ),
    );
    await faro.refreshFeatureFlags();

    expect(faro.isFeatureEnabled('new-ui', distinctId: 'user_1'), isTrue);
    // Llamarlo otra vez NO debe duplicar la exposición (dedup por variante).
    expect(faro.isFeatureEnabled('new-ui', distinctId: 'user_1'), isTrue);

    await faro.flush();
    await _waitFor(() => server.events.isNotEmpty);

    final exposures =
        server.events.where((e) => e['name'] == r'$feature_exposure').toList();
    expect(exposures, hasLength(1), reason: 'dedup: una sola exposición');
    final exp = exposures.first;
    expect(exp['type'], 'track');
    expect(exp['distinct_id'], 'user_1', reason: 'distinct_id override');
    final props = exp['properties'] as Map<String, dynamic>;
    expect(props['flag_key'], 'new-ui');
    expect(props['variant'], 'B');
    expect(props['enabled'], true);
  });

  // ---------- (c) conditions.properties no satisfechas → false sin exposición ----------

  test('conditions.properties no satisfechas → false y sin exposición',
      () async {
    server.flagsBody = {
      'project': 'demo',
      'flags': [
        {
          'key': 'beta',
          'rollout_percentage': 100,
          'conditions': {
            'properties': {'plan': 'pro'},
          },
        },
      ],
    };
    final faro = Faro.init(
      FaroOptions(
        endpoint: server.url,
        token: 'tk',
        service: 'ff-cond',
        installGlobalHandlers: false,
        flushInterval: const Duration(seconds: 100),
        featureFlagRefreshInterval: const Duration(seconds: 100),
      ),
    );
    await faro.refreshFeatureFlags();

    // plan != pro → no matchea → false, sin exposición.
    expect(
      faro.isFeatureEnabled(
        'beta',
        distinctId: 'user_1',
        properties: {'plan': 'free'},
      ),
      isFalse,
    );
    // Sanity: con el property correcto sí matchea (rollout 100 → true).
    expect(
      faro.isFeatureEnabled(
        'beta',
        distinctId: 'user_2',
        properties: {'plan': 'pro'},
      ),
      isTrue,
    );

    await faro.flush();
    await _waitFor(() => server.events.isNotEmpty);

    final exposures =
        server.events.where((e) => e['name'] == r'$feature_exposure').toList();
    expect(
      exposures,
      hasLength(1),
      reason: 'solo el caso que matchea genera exposición',
    );
    expect((exposures.first['properties'] as Map)['variant'], 'B');
    expect(exposures.first['distinct_id'], 'user_2');
  });

  // ---------- (d) close(timeout:) retorna aunque la red cuelgue ----------

  test('close(timeout:) no bloquea indefinidamente con la red colgada',
      () async {
    final faro = Faro.init(
      FaroOptions(
        endpoint: server.url,
        token: 'tk',
        service: 'close-timeout',
        installGlobalHandlers: false,
        flushInterval: const Duration(seconds: 100),
        featureFlagRefreshInterval: const Duration(seconds: 100),
        // httpTimeout alto: si close() respetara solo el http timeout, tardaría.
        httpTimeout: const Duration(seconds: 30),
      ),
    );
    server.hang = true; // a partir de aquí el server no responde nunca
    for (var i = 0; i < 5; i++) {
      faro.log(level: 'INFO', message: 'pendiente-$i');
    }

    final sw = Stopwatch()..start();
    await faro.close(timeout: const Duration(milliseconds: 300));
    sw.stop();

    expect(
      sw.elapsed,
      lessThan(const Duration(seconds: 5)),
      reason: 'close() debe cortar en ~timeout, no esperar al http timeout',
    );
  });

  // ---------- isFeatureEnabled sobre flag inexistente → false ----------

  test('flag desconocido → false', () async {
    final faro = Faro.init(
      FaroOptions(
        endpoint: server.url,
        token: 'tk',
        service: 'ff-unknown',
        installGlobalHandlers: false,
        flushInterval: const Duration(seconds: 100),
        featureFlagRefreshInterval: const Duration(seconds: 100),
      ),
    );
    expect(faro.isFeatureEnabled('nope', distinctId: 'x'), isFalse);
  });
}
